File: AlgAE/algae-cppserver/src/main/cpp/menuMessage.cpp
/*
 * menuMessage.cpp
 *
 *  Created on: Jul 11, 2012
 *      Author: zeil
 */


#include <algae/communications/menuMessage.h>
#include <algae/communications/xmlOutput.h>
#include <string>
#include <string>
#include <iostream>
#include <typeinfo>
#include <vector>

using namespace std;

namespace algae {

/**
 * Sent as an animation is starting, this message tells the client
 * what text should appear in the Help->About menu and what menu items
 * should be placed in the Algorithm menu listing functions that users
 * can select to see animations of.
 *
 * @author zeil
 *
 */


void MenuMessage::printXML (std::ostream& out) const
{
	JavaTag jt (out);
	{
		ObjectTag ot (out, communicationsPackage + ".MenuMessage");
		{
			PropertyTag pt (out, "about");
			{
				StringValue (out, about);
			}
		}
		{
			PropertyTag pt (out, "menuItems");
			{
				ArrayTag at (out, "java.lang.String", menuItems.size());
				for (vector<string>::size_type i = 0; i < menuItems.size(); ++i)
				{
					IndexTag vt(out, i);
					StringValue sv (out, menuItems[i]);
				}
				at.close();
			}
		}
	}
	jt.close();
}

ClientMessage* MenuMessage::clone() const
{
	return new MenuMessage(*this);
}

bool MenuMessage::operator== (const ClientMessage& other) const
{
	if (typeid(other) != typeid(MenuMessage))
		return false;
	else
	{
		const MenuMessage& msg =  (const MenuMessage&)other;
		return about == msg.about && menuItems == msg.menuItems;
	}
}

}

File: AlgAE/algae-client-server/src/main/java/edu/odu/cs/AlgAE/Common/Communications/MenuMessage.java
/**
 *
 */
package edu.odu.cs.AlgAE.Common.Communications;

import java.util.Arrays;

/**
 * Sent as an animation is starting, this message tells the client
 * what text should appear in the Help->About menu and what menu items
 * should be placed in the Algorithm menu listing functions that users
 * can select to see animations of.
 *
 * @author zeil
 *
 */
public class MenuMessage extends ClientMessage {
	
	private String about;
	private String[] menuItems;

	/**
	 * @param about text to appear in the Help menu
	 * @param menuItems list of strings to be used as menu items
	 */
	public MenuMessage(String about, String[] menuItems) {
		super("Menu");
		this.about = about;
		this.menuItems = Arrays.copyOf(menuItems, menuItems.length);
	}

	public MenuMessage() {
		super("Menu");
		this.about = "*No information provided*";
		this.menuItems = new String[0];
	}

	
	/**
	 * @return the help about string
	 */
	public String getAbout() {
		return about;
	}

	/**
	 * Set the help about string
	 */
	public void setAbout(String about) {
		this.about = about;
	}


	/**
	 * @return the menu items
	 */
	public String[] getMenuItems() {
		return menuItems;
	}

	/**
	 * @param menuItems the menu items to set
	 */
	public void setMenuItems(String[] menuItems) {
		this.menuItems = menuItems;
	}

	@Override
	public boolean equals(Object clientMessage) {
		if (clientMessage == null)
			return false;
		try {
			MenuMessage msg = (MenuMessage)clientMessage;
			return msg.about.equals(about)
					&& Arrays.equals(msg.menuItems, menuItems);
		} catch (Exception e) {
			return false;
		}
	}

}
